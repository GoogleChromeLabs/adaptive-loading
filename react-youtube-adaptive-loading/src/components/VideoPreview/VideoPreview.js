import React from 'react';
import {Image} from 'semantic-ui-react';
import './VideoPreview.scss';

import TimeAgo from 'javascript-time-ago';
import en from 'javascript-time-ago/locale/en';
import {getShortNumberString} from '../../services/number/number-format';
import {getVideoDurationString} from '../../services/date/date-format';
import {Link} from 'react-router-dom';

TimeAgo.locale(en);
const timeAgo = new TimeAgo('en-US');

export class VideoPreview extends React.Component {
  render() {
    const {video} = this.props;
    if (!video) {
      return <div/>;
    }

    const duration = video.contentDetails ? video.contentDetails.duration : null;
    const videoDuration = getVideoDurationString(duration);
    const viewAndTimeString = VideoPreview.getFormattedViewAndTime(video);
    const horizontal = this.props.horizontal ? 'horizontal' : null;
    const expanded = this.props.expanded ? 'expanded' : null;
    const description = this.props.expanded ? video.snippet.description : null;

    return (
      <Link to={{pathname: this.props.pathname, search: this.props.search}}>
        <div className={['video-preview', horizontal, expanded].join(' ')}>
          <div className='image-container'>
            <Image src={video.snippet.thumbnails.medium.url}/>
            <div className='time-label'>
              <span>{videoDuration}</span>
            </div>
          </div>

          <div className='video-info'>
            <div className={['semi-bold', 'show-max-two-lines', expanded].join(' ')}>{video.snippet.title}</div>
            <div className='video-preview-metadata-container'>
              <div className='channel-title'>{video.snippet.channelTitle}</div>
              <div className='view-and-time'>{viewAndTimeString}</div>
              <div className='show-max-two-lines'>{description}</div>
            </div>
          </div>
        </div>
      </Link>
    );
  }

  static getFormattedViewAndTime(video) {
    const publicationDate = new Date(video.snippet.publishedAt);
    const viewCount = video.statistics ? video.statistics.viewCount : null;
    if(viewCount) {
      const viewCountShort = getShortNumberString(video.statistics.viewCount);
      return `${viewCountShort} views â€¢ ${timeAgo.format(publicationDate)}`;
    }
    return '';
  }
}