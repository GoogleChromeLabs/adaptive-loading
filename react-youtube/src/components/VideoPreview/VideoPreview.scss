@import '../../styles/shared.scss';

.video-preview {
  display: grid;
  grid: 118px auto / 210px;

  /*
     override grid settings to make VideoPreview horizontal.
     Vertical is default
   */
  &.horizontal {
    grid: auto / 210px auto;

    &.expanded {
      grid: auto / 246px auto;
    }
    grid-column-gap: 4px;
    .video-info {
      grid-row: 1 / 2;
      grid-column: 2 / 3;
    }
  }
}

.video-info {
  grid-row: 2 / 3; /* is overriden when VideoPreview is horizontal */
  grid-column: 1 / 2; /* is overriden when VideoPreview is horizontal */
  color: $text-color-dark;

  .video-preview-metadata-container {
    padding-top: 5px;
    font-size: 13px;
    color: #6E6E6E;

    .channel-title {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  }

  .show-max-two-lines {
    overflow: hidden;
    line-height: 1.4em;
    max-height: 2.8em;
  }

  .view-and-time {
    margin-bottom: 10px;
  }

  .semi-bold {
    &.expanded {
      font-weight: 400;
      font-size: 1.3rem;
    }
    font-weight: 600;
  }
}

.image-container {
  position: relative;
  grid-row: 1 / 2;
  grid-column: 1 / 2;

  /* Video duration label at bottom right */
  .time-label {
    position: absolute;
    background: $text-color-dark;
    bottom: 0;
    right: 0;
    opacity: 0.8;
    border-radius: 2px;
    font-weight: 500;
    color: white;
    margin: 4px;
    padding: 2px 4px;
    line-height: 12px;
  }
}
