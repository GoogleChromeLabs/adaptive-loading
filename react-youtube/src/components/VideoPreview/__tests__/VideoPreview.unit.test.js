import React from 'react';
import {shallow} from 'enzyme';
import {VideoPreview} from '../VideoPreview';

describe('VideoPreview', () => {
  test('renders', () => {
    const wrapper = shallow(<VideoPreview/>);
    expect(wrapper).toMatchSnapshot();
  });
});