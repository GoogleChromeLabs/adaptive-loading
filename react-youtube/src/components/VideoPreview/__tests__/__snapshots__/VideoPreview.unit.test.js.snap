// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview renders 1`] = `
<div
  className="video-preview"
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="http://via.placeholder.com/210x118"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span>
        05:22
      </span>
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      Video title
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        Channel title
      </div>
      <div>
        <span>
          42.0B views â€¢ 2 days ago
        </span>
      </div>
    </div>
  </div>
</div>
`;
